Metrics

ejholmes opened the pull request:
This sets up a small stats framework, and adds some middleware to all AWS clients that will increment some counters for aws requests and responses. This will allow us to get better visibility into throttling, or other issues.

We can also start using the `stats.*` methods to generate metrics for the other things, like number of deploys, how long deploys take, etc.

The stats package is just a small layer of abstraction over a statsd'ish client, and I've only provided a datadog implementation for now.

**Metrics we should have**

* [x] `aws.request.*` For outbound AWS requests
* [x] `heroku.request.*` For inbound Heroku API requests.
* [x] `runtime.*` Golang runtime stats.
* [ ] `heroku.auth.error` for authorization errors.
* [ ] `core.*` for core API calls.
* [ ] `scheduler.cloudformation.template_size` For tracking generated template sizes

phobologic commented:
Would be cool if we had an `errors` metric that we could tag with different types of errors/urls, etc.

ejholmes commented:
Good call, I can add that to the embedded error reporter 👍

phobologic commented on cmd/empire/factories.go:
```diff
@@ -418,13 +380,34 @@ func newHBReporter(key, env string) (reporter.Reporter, error) {
+	return stats.Null, nil
+}
+
+func newDataDogStats(addr string) (stats.Stats, error) {
```
It might be worth at least having a datadog & plain statsd implementation at first.  I think the only difference would be removing tags?

ejholmes replied:
Cool, yeah, I can add a plain statsd backend as well.

phobologic commented on cmd/empire/factories.go:
```diff
@@ -445,3 +428,48 @@ func newAuthProvider(c *cli.Context) (dockerauth.AuthProvider, error) {
+					tags = append(tags, fmt.Sprintf("error:%s", err.Code()))
+					stats.Inc(fmt.Sprintf("aws.request.error"), 1, 1.0, tags)
+					stats.Inc(fmt.Sprintf("aws.request.%s.error", r.ClientInfo.ServiceName), 1, 1.0, tags)
+					stats.Inc(fmt.Sprintf("aws.request.%s.%s.error", r.ClientInfo.ServiceName, r.Operation.Name), 1, 1.0, tags)
```
nm on the error stat :) sorry for jumping the gun

phobologic commented:
So awesome 👍

ejholmes commented:
Added a commit that adds a statsd backend as well, and also changed the flag to `--stats`, so you can configure it like this:

```console
empire server --stats=statsd://localhost:8125
```

Or

```console
empire server --stats=dogstatsd://localhost:8126
```

ejholmes commented:
Also added some sample of the go runtime metrics every 30 seconds as well.
